cloud/libvirt/actuators/machine/utils/network.go

```diff
@@ -0,0 +1,319 @@
+	log.Printf("Adding host with XML:\n%s", xmlDesc)
+	return n.Update(libvirt.NETWORK_UPDATE_COMMAND_ADD_LAST, libvirt.NETWORK_SECTION_IP_DHCP_HOST, -1, xmlDesc, libvirt.NETWORK_UPDATE_AFFECT_CURRENT)
+}
+
+func getHostXMLDesc(ip, mac, name string) string {
+	dd := libvirtxml.NetworkDHCPHost{
+		IP:   ip,
+		MAC:  mac,
+		Name: name,
+	}
+	tmp := struct {
+		XMLName xml.Name `xml:"host"`
+		libvirtxml.NetworkDHCPHost
+	}{xml.Name{}, dd}
+	xml, err := xmlMarshallIndented(tmp)
+	if err != nil {
+		panic("could not marshall host")
+	}
+	return xml
+}
+
+// Update a static host from the network
+func updateHost(n *libvirt.Network, ip, mac, name string) error {
+	xmlDesc := getHostXMLDesc(ip, mac, name)
+	log.Printf("Updating host with XML:\n%s", xmlDesc)
+	return n.Update(libvirt.NETWORK_UPDATE_COMMAND_MODIFY, libvirt.NETWORK_SECTION_IP_DHCP_HOST, -1, xmlDesc, libvirt.NETWORK_UPDATE_AFFECT_CURRENT)
+}
+
+// randomMACAddress returns a randomized MAC address
+func randomMACAddress() (string, error) {
+	buf := make([]byte, 6)
+	_, err := rand.Read(buf)
+	if err != nil {
+		return "", err
+	}
+
+	// set local bit and unicast
+	buf[0] = (buf[0] | 2) & 0xfe
+	// Set the local bit
+	buf[0] |= 2
+
+	// avoid libvirt-reserved addresses
+	if buf[0] == 0xfe {
+		buf[0] = 0xee
+	}
+
+	return fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x",
+		buf[0], buf[1], buf[2], buf[3], buf[4], buf[5]), nil
+}
+
+// Creates a network definition from a XML
+func newDefNetworkFromXML(s string) (libvirtxml.Network, error) {
+	var networkDef libvirtxml.Network
+	err := xml.Unmarshal([]byte(s), &networkDef)
+	if err != nil {
+		return libvirtxml.Network{}, err
+	}
+	return networkDef, nil
+}
+
+// Creates a network definition with the defaults the provider uses
+func newNetworkDef() libvirtxml.Network {
+	const defNetworkXML = `
+		<network>
+		  <name>default</name>
+		  <forward mode='nat'>
+		    <nat>
+		      <port start='1024' end='65535'/>
+		    </nat>
+		  </forward>
+		</network>`
+	if d, err := newDefNetworkFromXML(defNetworkXML); err != nil {
+		panic(fmt.Sprintf("Unexpected error while parsing default network definition: %s", err))
+	} else {
+		return d
+	}
+}
+
+func CreateNetwork(name, domain, bridge, mode string, addresses []string, autostart bool, client *Client) error {
+	// see https://libvirt.org/formatnetwork.html
+	virConn := client.libvirt
+
+	networkDef := newNetworkDef()
+	networkDef.Name = name
+	networkDef.Domain = &libvirtxml.NetworkDomain{
+		Name: domain,
+	}
+
+	// use a bridge provided by the user, or create one otherwise (libvirt will assign on automatically when empty)
+	b := ""
+	if bridge != "" {
+		b = bridge
+	}
+	networkDef.Bridge = &libvirtxml.NetworkBridge{
+		Name: b,
+		STP:  "on",
+	}
+
+	// check the network mode
+	networkDef.Forward = &libvirtxml.NetworkForward{
+		Mode: mode,
+	}
+	if networkDef.Forward.Mode == netModeIsolated || networkDef.Forward.Mode == netModeNat || networkDef.Forward.Mode == netModeRoute {
```

Comment:
All of this is difficult to read; could it be split up into some helper functions?

Reply:
this is mainly decoupling utils so we have a real DSL for libvirt and we can move it to pkg, so we can raise the bar and we can have subsequent PRs clean and tested, let's do farther logic improvements in follow ups